### ISSUE TITLE
multiple bugs in clusteringAnalysis

### ISSUE THREAD

**Issue by mathildelpx** [oliviermirat/ZebraZoom#26]
Hi Olivier,

I found multiple bugs when trying to run clusteringAnalysis.

- OS: Linux Mint
- version: zebrazoom 1.32.5
- installation process: pip install zebrazoom
- IDE: command prompt

**reproduce**

``` bash

python -m zebrazoom dataPostProcessing clusteringAnalysis /.../SCAPE_data_220222.xls 0 5 3 KMeans 1

Traceback (most recent call last):
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/__main__.py", line 163, in <module>
    clusteringAnalysis(sys)
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/clusteringAnalysis.py", line 44, in clusteringAnalysis
    generatePklDataFileForVideo(os.path.join(os.path.split(pathToExcelFile)[0], nameWithExt), os.path.join(cur_dir_path, 'ZZoutput'), frameStepForDistanceCalculation)
NameError: name 'frameStepForDistanceCalculation' is not defined


```

**Proposed solution:**

This error comes from the fact that the parameter  'frameStepForDistanceCalculation'  is expected to be defined in clusteringAnalysis.py to launch the dataframe creation. I fixed this by defining the parameter in the script as equal to 5. I don't know what this parameter is for though.

## Other bug

Then I launched again and got a new error: 

```
Creating pandas dataframe of parameters
Parameter: frameStepForDistanceCalculation: 5
Calculating and storing all parameters:
Traceback (most recent call last):
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/__main__.py", line 163, in <module>
    clusteringAnalysis(sys)
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/clusteringAnalysis.py", line 46, in clusteringAnalysis
    generatePklDataFileForVideo(os.path.join(os.path.split(pathToExcelFile)[0], nameWithExt), os.path.join(cur_dir_path, 'ZZoutput'), frameStepForDistanceCalculation)
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/dataAnalysis/datasetcreation/generatePklDataFileForVideo.py", line 28, in generatePklDataFileForVideo
    createPandasDataFrameOfParameters(hyperparameters, videoName, videoExtension, ZZoutputLocation)
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/code/dataPostProcessing/createPandasDataFrameOfParameters.py", line 47, in createPandasDataFrameOfParameters
    [conditions, genotypes, nbFramesTakenIntoAccount, globParam] = createDataFrame(dataframeOptions, excelFileDataFrame, 1, ['BoutFrameNumberStart', 'tailAngleSymmetry', 'secondBendAmpDividedByFirst', 'tailAngleIntegral'])
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/dataAnalysis/datasetcreation/createDataFrame.py", line 159, in createDataFrame
    with open(os.path.join(path, 'results_' + trial_id + '.txt')) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/ZZoutput/210819_F1_runA_run1/results_210819_F1_runA_run1.txt'
```

HOWEVER, it is not looking at the path I defined for this file in the xls file. It is taking the ZZ default path, so of course there is not zebrazoom output file in this folder. I attached the experiment file.
[SCAPE_data_220222.xls](https://github.com/oliviermirat/ZebraZoom/files/8126682/SCAPE_data_220222.xls)
.

**Comment by oliviermirat** [comment]
Hi Mathilde,
Thanks for the feedback, I will look into this asap, but please patient, I'm overloaded with work at the moment (hopefully I can look into this on Friday).

**Comment by mathildelpx** [comment]
Hi Olivier,

Of course, no worries! I can always use previous version (1.30.21) where this function works.

**Comment by oliviermirat** [comment]
I just released a new version that should solve this issue. Can you please check if it works for you now?

I also updated the documentation about this: https://zebrazoom.org/documentation/docs/behaviorAnalysis/behaviorAnalysisThroughCommandLine/#clustering-analysis-for-zebrafish-only-through-the-command-line-instead-of-from-the-gui

**Comment by mathildelpx** [comment]
Hi Olivier,

cool, thanks a lot. Previous issue were fixed indeed, I however ran into another bug related to number of frames taken into account.

```
aceback (most recent call last):
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/runpy.py", line 194, in _run_module_as_main
    return _run_code(code, main_globals, None,
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/__main__.py", line 163, in <module>
    clusteringAnalysis(sys)
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/clusteringAnalysis.py", line 48, in clusteringAnalysis
    [conditions, genotypes, nbFramesTakenIntoAccount, globParam] = createDataFrame(dataframeOptions, "", 0, ['BoutFrameNumberStart', 'tailAngleSymmetry', 'secondBendAmpDividedByFirst', 'tailAngleIntegral'])
  File "/home/mathilde.lapoix/anaconda3/envs/ZebraZoom/lib/python3.8/site-packages/zebrazoom/dataAnalysis/datasetcreation/createDataFrame.py", line 171, in createDataFrame
    raise ValueError("nbFramesTakenIntoAccount was too low when pre-generating the pkl file of a video:" + str(np.sum(['instaTBF' in param for param in dfReloadedVid.columns.tolist()])) + " , " + str(np.sum(['tailAnglesRecalculated' in param for param in dfReloadedVid.columns.tolist()])) + " , " + str(np.sum(['instaSpeed' in param for param in dfReloadedVid.columns.tolist()])) + " ; nbFramesTakenIntoAccount :" + str(nbFramesTakenIntoAccount))
ValueError: nbFramesTakenIntoAccount was too low when pre-generating the pkl file of a video:100 , 0 , 100 ; nbFramesTakenIntoAccount :151
```

This issue was fixed when I manually changed clusteringAnalysis.py to put 70 instead of -1.


Related to that, I was thinking it could be nice to add this parameter as one the user could set from the command line. I think the values need to be quite different depending on the conditions and the frame rate of acquisition. For example on my side, I use 70 frames (~230ms) because my bouts are longer in head-embedded conditions. (However, I didn't test how much frames were explaining how much of the variance between my bouts). 

This issue can wait, I'm just reporting it but it does not prevent me from using it.

**Comment by oliviermirat** [comment]
I need to add some explanations about all this in the documentation but briefly, when nbFramesTakenIntoAccount is set to -1, it is automatically re-calculated by taking the median value of the number of frames of the first 100 bouts in the dataset.

Also, I very recently introduced a new feature that stores the kinematic parameters and the clustering parameters when they are calculated for the first time and then reloads them afterwards when they are needed for future analysis. However, when I first introduced that new feature (about 2 weeks ago) I forgot that nbFramesTakenIntoAccount could change from one analysis to the next leading to some dataset having smaller nbFramesTakenIntoAccount values than others. I've now fixed that by always taking a very large number for nbFramesTakenIntoAccount when storing the parameters, and then down-sampling after reloading them in future analysis. In the error you are reporting here, my guess is that parameters are automatically being reloaded with inconsistent values for nbFramesTakenIntoAccount calculated with the previous ZZ release which I think explains the bug you are seeing.

This explanation is probably a bit complicated to follow but long story short if you remove the files videoName.pkl and frameStepForDistanceCalculationUsed.txt that should be stored in the tracking result folder and relaunched the analysis with nbFramesTakenIntoAccount  at -1 then it should work (and most likely the parameter nbFramesTakenIntoAccount  will automatically be set at a value higher than 70).
